fem/nonlininteg.cpp

```diff
@@ -813,4 +816,121 @@ void VectorConvectionNLFIntegrator::AssembleElementGrad(
    }
 }
 
+
+void ConvectiveVectorConvectionNLFIntegrator::AssembleElementGrad(
+   const FiniteElement &el,
+   ElementTransformation &trans,
+   const Vector &elfun,
+   DenseMatrix &elmat)
+{
+   int nd = el.GetDof();
+   int dim = el.GetDim();
+
+   shape.SetSize(nd);
+   dshape.SetSize(nd, dim);
+   dshapex.SetSize(nd, dim);
+   elmat.SetSize(nd * dim);
+   elmat_comp.SetSize(nd);
+   gradEF.SetSize(dim);
+
+   EF.UseExternalData(elfun.GetData(), nd, dim);
+
+   double w;
+   Vector vec1(dim), vec2(dim), vec3(nd);
+
+   const IntegrationRule *ir = IntRule ? IntRule : &GetRule(el, trans);
+
+   elmat = 0.0;
+   for (int i = 0; i < ir->GetNPoints(); i++)
+   {
+      const IntegrationPoint &ip = ir->IntPoint(i);
+      trans.SetIntPoint(&ip);
+
+      el.CalcShape(ip, shape);
+      el.CalcDShape(ip, dshape);
+
+      w = ip.weight;
+
+      if (Q)
+      {
+         w *= Q->Eval(trans, ip);
+      }
+
+      EF.MultTranspose(shape, vec1); // u^n
+
+      trans.AdjugateJacobian().Mult(vec1, vec2);
+
+      vec2 *= w;
+      dshape.Mult(vec2, vec3); // (u^n \cdot grad u^{n+1})
+      MultVWt(shape, vec3, elmat_comp); // (u^n \cdot grad u^{n+1},v)
+
+      for (int i = 0; i < dim; i++)
+      {
+         elmat.AddMatrix(elmat_comp, i * nd, i * nd);
+      }
+   }
+}
+
+
+void SkewSymmetricVectorConvectionNLFIntegrator::AssembleElementGrad(
+   const FiniteElement &el,
+   ElementTransformation &trans,
+   const Vector &elfun,
+   DenseMatrix &elmat)
+{
+   int nd = el.GetDof();
+   int dim = el.GetDim();
+
+   shape.SetSize(nd);
+   dshape.SetSize(nd, dim);
+   dshapex.SetSize(nd, dim);
+   elmat.SetSize(nd * dim);
+   elmat_comp.SetSize(nd);
+   gradEF.SetSize(dim);
+
+   DenseMatrix elmat_comp_T(nd);
+
+   EF.UseExternalData(elfun.GetData(), nd, dim);
+
+   double w;
```

Comment:
I would have this declared within the loop. You could even make use of a ternary operation internally to make it a `const double` as well by doing something like:

```c++
const double w = Q ? Q->Eval(trans, ip) * ip.weight : ip.weight;
```

Although, I would say that sacrifices readability of the code a bit.

Reply:
The implementation is based on VectorConvectionNLFIntegrator::AssembleElementGrad, which I did not write. I'm open to making this change, however I think it should be consistent with VectorConvectionNLFIntegrator::AssembleElementGrad, which I think is beyond the scope of this PR (even if it's a relatively minor change).

Reply:
Since no one seemed to have an issue with it I went ahead and just made this suggested change.